homeassistant/components/proximity/coordinator.py

```diff
@@ -0,0 +1,246 @@
+"""Data update coordinator for the Proximity integration."""
+
+import logging
+from typing import Any
+
+from homeassistant.const import (
+    ATTR_LATITUDE,
+    ATTR_LONGITUDE,
+    CONF_DEVICES,
+    CONF_UNIT_OF_MEASUREMENT,
+    CONF_ZONE,
+    UnitOfLength,
+)
+from homeassistant.core import HomeAssistant, State
+from homeassistant.helpers.event import async_track_state_change
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+from homeassistant.util.location import distance
+from homeassistant.util.unit_conversion import DistanceConverter
+
+from .const import (
+    ATTR_DIR_OF_TRAVEL,
+    ATTR_DIST_TO,
+    ATTR_NEAREST,
+    CONF_IGNORED_ZONES,
+    CONF_TOLERANCE,
+    DEFAULT_DIR_OF_TRAVEL,
+    DEFAULT_DIST_TO_ZONE,
+    DEFAULT_NEAREST,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ProximityDataUpdateCoordinator(
+    DataUpdateCoordinator[dict[str, str | int | float]]
+):
+    """Proximity data update coordinator."""
+
+    def __init__(
+        self, hass: HomeAssistant, friendly_name: str, config: ConfigType
+    ) -> None:
+        """Initialize the Proximity coordinator."""
+        self.ignored_zones: list[str] = config[CONF_IGNORED_ZONES]
+        self.proximity_devices: list[str] = config[CONF_DEVICES]
+        self.tolerance: int = config[CONF_TOLERANCE]
+        self.proximity_zone: str = config[CONF_ZONE]
+        self.unit_of_measurement: str = config.get(
+            CONF_UNIT_OF_MEASUREMENT, hass.config.units.length_unit
+        )
+        self.friendly_name = friendly_name
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=friendly_name,
+            update_interval=None,
+        )
+
+        self.data = {
+            ATTR_DIST_TO: DEFAULT_DIST_TO_ZONE,
+            ATTR_DIR_OF_TRAVEL: DEFAULT_DIR_OF_TRAVEL,
+            ATTR_NEAREST: DEFAULT_NEAREST,
+        }
+
+        self.old_state: State | None = None
+        self.new_state: State | None = None
+        self.entity: str
+        async_track_state_change(
+            hass, self.proximity_devices, self.async_check_proximity_state_change
+        )
+
+    async def async_check_proximity_state_change(
+        self, entity: str, old_state: State | None, new_state: State | None
+    ) -> None:
+        """Fetch and process state change event."""
+        self.old_state = old_state
+        self.new_state = new_state
+        self.entity = entity
+        await self.async_refresh()
+
+    async def _async_update_data(self) -> dict[str, Any]:
```

Comment:
Consider defining a dataclass with the 3 fields as the return value instead. It can help enforce types (like elsewhere I saw there were some calls to convert the return values to strings) and it can handle the default values cases.

Reply:
i already tried to define an own dataclass for this, but TBH did not get it work nor get mypy happy 🙈 
but yes, this should be done ... will try again

Reply:
--> 005ba6d